Server side annotation aggregation

dihydrogenmonoxide16 opened the pull request:
Proposal describes the way how clients and servers could interact in a more efficient and scalable way working with annotation (reaction) - heavy rooms.

rendered (TBD)

turt2live commented:
@dihydrogenmonoxide16 when you get a chance, please [sign off](https://github.com/matrix-org/matrix-spec-proposals/blob/main/CONTRIBUTING.md#sign-off) on the changes for this MSC to be eligible for FCP acceptance. Also, lines should be limited to approximately 120 characters.

dihydrogenmonoxide16 commented:
> @dihydrogenmonoxide16 when you get a chance, please [sign off](https://github.com/matrix-org/matrix-spec-proposals/blob/main/CONTRIBUTING.md#sign-off) on the changes for this MSC to be eligible for FCP acceptance. Also, lines should be limited to approximately 120 characters.

addressed

dihydrogenmonoxide16 commented:
@turt2live , I added changes to make the behavior backwards-compatible with older clients as we discussed before

turt2live commented on proposals/4074-server-side-annotations-aggregation.md:
```diff
@@ -0,0 +1,270 @@
+Servers tried to provide clients with server side generated
+annotation aggregates before, but these attempts were not successful
+mainly because both sides (server and client) tried to
+aggregate and the same time. That used to lead to the situation when
```
and *what* at the same time?

turt2live commented on proposals/4074-server-side-annotations-aggregation.md:
```diff
@@ -0,0 +1,270 @@
+
+Clients will not be responsible for aggregation of relations of
+"m.annotation" relation type anymore. Such relations will always be
+aggregated by the server. The only exception to this rule is E2E
+encrypted events, which should be solely aggregated by the client.
```
Does the encryption condition only apply once [reactions are encrypted](https://github.com/matrix-org/matrix-spec/issues/660)?

turt2live commented on proposals/4074-server-side-annotations-aggregation.md:
```diff
@@ -0,0 +1,270 @@
+[`relations`](https://spec.matrix.org/v1.8/client-server-api/#relationships-api), [`get event by id`](https://spec.matrix.org/v1.8/client-server-api/#get_matrixclientv3roomsroomideventeventid) and
+similar client api endpoints.
+
+For server-server API, homeserver should provide aggregates,
+including only local events (events created and signed by self).
```
I'm not following this bit. Servers distribute the events as-is without consideration for the underlying feature, meaning each server should be able to determine how many of each event there is. Reactions can also be used as a form of "DAG healing" when rooms get complex. This is because they are normally higher traffic than regular messages, giving more opportunity for the DAG to heal extremities. The delay associated with needing to process these complicated event trees is also (typically) less critical than message traffic.

turt2live commented on proposals/4074-server-side-annotations-aggregation.md:
```diff
@@ -0,0 +1,270 @@
+For server-server API, homeserver should provide aggregates,
+including only local events (events created and signed by self).
+
+Servers are allowed to limit maximum number of different "m.annotation"
+keys they aggregate to a single event by a reasonable value (that
+corresponds to a limited number of different reaction kinds, e.g. "👍",
+"👎"). This is needed to avoid the situation when malicious users may
+attack server creating jumbo-events (events with arbitrary high count of
+different reaction kinds). Servers may configure this threshold on their
+discretion, but this number should not be lower than 16. When annotation
+key limit is reached, no new keys should be added to the aggregate, but
+older keys would continue to be aggregated (change their count).
+Annotations beyond that threshold should still be available via
+[`relations`](https://spec.matrix.org/v1.8/client-server-api/#relationships-api) endpoints.
```
How would a server communicate that it hit the limit?

dihydrogenmonoxide16 replied:
I completely removed this clause. Potentially this problem already exists now, but only on the client side. I think it would better get addressed separately in order not to overcomplicate this change proposal (already big enough and technically complex)

turt2live commented on proposals/4074-server-side-annotations-aggregation.md:
```diff
@@ -0,0 +1,270 @@
+
+## Unstable prefix
+
+No new identifiers are proposed; it is proposed that servers implementing
+this
+proposal simply do so on the existing endpoints.
```
`filter_server_aggregated_relation_types` and the `m.reaction` EDU appear new, as do a few other identifiers. They will need to be namespaced.

dihydrogenmonoxide16 replied:
prefixed with `msc4074.`, also renamed `filter_server_aggregated_relation_types` to `msc4072.not_aggregated_relations` to better match other filter params naming

turt2live reviewed (approved):
overall this seems like a great concept. I'm not sure on the client synchronization stuff myself, but others should be. A listed client implementation would help quite a bit in this understanding, I suspect.

turt2live commented on proposals/4074-server-side-annotations-aggregation.md:
Implementation requirements:

* [ ] Server supporting filter opt-in behaviour
* [ ] Client using the filter behaviour

dihydrogenmonoxide16 replied:
What is the expectation here?

- to add this comment to the MXC inself
- some pseudocode
- working client/server implementation somewhere
- arch doc
?

Or is it just a ref comment for other contributors?

tulir replied:
The third option, a working client/server implementation somewhere. MSCs must be proven with implementations before they can be accepted: https://spec.matrix.org/proposals/#implementing-a-proposal

It generally doesn't matter who implements it or where, so implementation requirements are commented as a reference for everyone

dihydrogenmonoxide16 commented:
> overall this seems like a great concept. I'm not sure on the client synchronization stuff myself, but others should be. A listed client implementation would help quite a bit in this understanding, I suspect.

@turt2live , I'm doing a follow-up work to answer your comments and extend/adjust the proposal. Will post updates in ~1 week due to high work load.

turt2live commented:
No worries! Take the time you need :)

dihydrogenmonoxide16 commented:
> No worries! Take the time you need :)

@turt2live, thank you for your feedback and collaboration! I answered your comments and addressed the points you mentioned. Added few improvements that turned to be necessary for a few niche corner cases (gap resolution, older event updates). Asked few questions there also.
